src/communication/receivers.rs

```diff
@@ -86,3 +93,53 @@ pub async fn run_receivers(mut receivers: Vec<ERDOSReceiver>) -> Result<(), Comm
     future::join_all(receivers.iter_mut().map(|receiver| receiver.run())).await;
     Ok(())
 }
+
+/// Listens on a TCP stream, and pushes control messages it receives to the node.
+#[allow(dead_code)]
+pub struct ControlReceiver {
+    /// The id of the node the stream is receiving data from.
+    node_id: NodeId,
+    /// Framed TCP read stream.
+    stream: SplitStream<Framed<TcpStream, ControlMessageCodec>>,
+    /// Mapping between stream id to pushers.
+    channel_to_handler: UnboundedSender<ControlMessage>,
+}
+
+impl ControlReceiver {
+    pub fn new(
+        node_id: NodeId,
+        stream: SplitStream<Framed<TcpStream, ControlMessageCodec>>,
+        channel_to_handler: UnboundedSender<ControlMessage>,
+    ) -> Self {
+        Self {
+            node_id,
+            stream,
+            channel_to_handler,
+        }
+    }
+
+    pub async fn run(&mut self) -> Result<(), CommunicationError> {
+        while let Some(res) = self.stream.next().await {
+            match res {
+                Ok(msg) => {
+                    self.channel_to_handler
```

Comment:
Should we update channel_to_handler regularly to ensure that we have up-to-date mappings between channels and handlers?